packages/cojson-transport-ws/src/index.ts

```diff
@@ -29,72 +47,73 @@ export function createWebSocketPeer(options: {
             >();
         const outgoing = yield* Queue.unbounded<SyncMessage>();
 
-        ws.addEventListener("close", (event) => {
-            void Effect.runPromiseExit(
-                Queue.offer(
-                    incoming,
-                    Either.left(
-                        new DisconnectedError(`${event.code}: ${event.reason}`),
-                    ),
+        yield* addEventListener(ws, "close", (event) =>
+            Queue.offer(
+                incoming,
+                Either.left(
+                    new DisconnectedError(`${event.code}: ${event.reason}`),
                 ),
-            ).then((e) => {
-                if (Exit.isFailure(e) && !Exit.isInterrupted(e)) {
-                    console.warn("Failed closing ws", e);
-                }
-            });
-        });
+            ).pipe(
+                Effect.onExit((e) => {
+                    if (Exit.isFailure(e) || Exit.isInterrupted(e)) {
+                        return Console.warn("Failed closing ws", e);
+                    }
+                    return Effect.void;
+                }),
+            ),
+        );
 
-        let pingTimeout: ReturnType<typeof setTimeout> | undefined;
+        let pingTimeout: Fiber.Fiber<void> | undefined;
 
-        ws.addEventListener("message", (event) => {
-            const msg = JSON.parse(event.data as string);
+        yield* addEventListener(ws, "message", (event) =>
+            Effect.gen(function* () {
+                const msg = yield* Effect.sync(() =>
+                    JSON.parse(event.data as string),
+                );
```

Comment:
I wonder if we should use Schema here to make sure it's a valid struct, or at least a valid JSON (`Schema.parseJson(Schema.Unknown)`)